tools/build_langserver/langserver/analyzer.go

```diff
@@ -1,3 +1,203 @@
 package langserver
 
+import (
+	"core"
+	"parse"
+	"parse/rules"
+	"sort"
+	"strconv"
+	"strings"
+
+	"fmt"
+	"io/ioutil"
+	"parse/asp"
+	"tools/build_langserver/lsp"
+)
+
 // TODO(bnmetrics): This file should contain functions to retrieve builtin and custom definitions of build defs
+
+// Analyzer is a wrapper around asp.parser
+// This is being loaded into a handler on initialization
+type Analyzer struct {
+	parser   *asp.Parser
+	BuiltIns map[string]*RuleDef
+}
+
+// RuleDef is a wrapper around asp.FuncDef,
+// it also includes a Header(function definition)
+// And Argument map stores the name and the information of the arguments this rule has
+type RuleDef struct {
+	*asp.FuncDef
+	Header string
+	ArgMap map[string]*Argument
+}
+
+// Argument is a wrapper around asp.Argument,
+// this is used to store the argument information for specific rules,
+// and it also tells you if the argument is required
+type Argument struct {
+	*asp.Argument
+	definition string
+	required   bool
+}
+
+// Identifier is a wrapper around asp.Identifier
+// Including the starting line and the ending line number
+type Identifier struct {
+	*asp.IdentStatement
+	Type      string
+	StartLine int
+	EndLine   int
+}
+
+func newAnalyzer() *Analyzer {
+	state := core.NewDefaultBuildState()
+	parser := parse.NewAspParser(state)
+
+	a := &Analyzer{
+		parser: parser,
+	}
+	a.builtInsRules()
+
+	return a
+}
+
+// BuiltInsRules gets all the builtin functions and rules as a map, and store it in Analyzer.BuiltIns
+// This is typically called when instantiate a new Analyzer
+func (a *Analyzer) builtInsRules() error {
+	statementMap := make(map[string]*RuleDef)
+
+	dir, _ := rules.AssetDir("")
+	sort.Strings(dir)
+	// Iterate through the directory and get the builtin statements
+	for _, filename := range dir {
+		if !strings.HasSuffix(filename, ".gob") {
+			asset := rules.MustAsset(filename)
+			stmts, err := a.parser.ParseData(asset, filename)
+			if err != nil {
+				log.Fatalf("%s", err)
+			}
+			// Iterate through the statement we got and add to statementMap
+			for _, statement := range stmts {
+				if statement.FuncDef != nil && !strings.HasPrefix(statement.FuncDef.Name, "_") {
+					content := string(asset)
+					statementMap[statement.FuncDef.Name] = newRuleDef(content, statement)
+				}
+			}
+		}
+	}
+
+	a.BuiltIns = statementMap
+	return nil
+}
+
+// IdentFromPos gets the Identifier given a lsp.Position
+func (a *Analyzer) IdentFromPos(uri lsp.DocumentURI, position lsp.Position) (*Identifier, error) {
+	idents, err := a.IdentFromFile(uri)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, ident := range idents {
+		if position.Line >= ident.StartLine && position.Line <= ident.EndLine {
```

Comment:
do you not need to check columns here as well?

Reply:
No, not at this stage, because I would need to get the whole identifier, like a function call(build target), then the detailed analysis to specific column is done at the hover/completion level